### ISSUE TITLE
`get_reserve_proof` returns valid proof with amount greater than account balance

### ISSUE THREAD

**Issue by woodser** [monero-project/monero#6595]
With unconfirmed transactions, `get_reserve_proof` returns a valid proof when called with an account index and an amount greater than the account's balance.  Checking the proof gives a total amount greater than the requested proof amount and a non-zero `spent` amount.  I figured the proof would only succeed with an amount less than or equal to the account balance?

**Comment by moneromooo-monero** [comment]
Seems to work for me. I added logs, and:

curl -k -X POST $URL -d '{"jsonrpc":"2.0","id":"0","method":"get_reserve_proof","params":{"account_index":1,"amount":"13529939118000","message":"Test message"}}' -H 'Content-Type: application/json'
{
  "error": {
    "code": -1,
    "message": "Not enough balance in this account for the requested minimum reserve amount"
  },
  "id": "0",
  "jsonrpc": "2.0"
}

2020-05-28 11:03:27.275	I get_reserve_proof: account_minreserve is set
2020-05-28 11:03:27.275	I    = 1 13.529939118000
2020-05-28 11:03:27.277	I Found usable output 6019, account 1, amount 0.899818240000
2020-05-28 11:03:27.277	I Found usable output 6021, account 1, amount 5.000000000000
2020-05-28 11:03:27.277	I Found usable output 6023, account 1, amount 5.000000000000
2020-05-28 11:03:27.277	I Total amount found: 10.899818240000
2020-05-28 11:03:27.278	E account_minreserve && balance < account_minreserve->second. THROW EXCEPTION: error::wallet_internal_error
2020-05-28 11:03:27.278	W /src/wallet/wallet2.cpp:11813:N5tools5error21wallet_internal_errorE: Not enough balance in this account for the requested minimum reserve amount

**Comment by moneromooo-monero** [comment]
Please send the JSON you're sending, and the balance of the account in question.

**Comment by woodser** [comment]
`get_reserve_proof` is called with this JSON: `{"method":"get_reserve_proof","id":"0","jsonrpc":"2.0","params":{"amount":"41900886994506","account_index":0,"message":"Test message"}}`

`get_accounts` returns `...{account_index=0, balance=41825886994506, ...}`

A proof signature is returned.  Calling `check_reserve_proof` with the signature returns:

`{id=0, jsonrpc=2.0, result={good=true, spent=29169365936882, total=42370606771561}}`

Account 0 has an unconfirmed incoming transfer (from a different account in the same wallet).

**Comment by moneromooo-monero** [comment]
Is this a testnet/stagenet wallet you could send me the keys to ?

**Comment by woodser** [comment]
Yes, it's a stagenet wallet:

`hijack lucky rally sober hockey robot gumball amaze gave fifteen organs gecko skater wizard demonstrate upright system vegan tobacco tsunami lurk withdrawn tomorrow uphill organs`

Height of first transaction: 589429

**Comment by moneromooo-monero** [comment]
That JSON gets me "Not enough balance in this account for the requested minimum reserve amount". I guess I fixed it in my work branch. I did make changes to the reserve proof stuff for historical proofs, but that relies on some database changes so it'll be a bit.

**Comment by woodser** [comment]
That JSON will give me the same thing unless there's an unconfirmed incoming transfer to the account.

**Comment by moneromooo-monero** [comment]
Ah. Do you know what height your daemon was when you ran this and it failed ?

**Comment by woodser** [comment]
No, but I just re-ran and reproduced the issue (height 592428).

**Comment by moneromooo-monero** [comment]
Still works at that height, with a tx in the txpool. You may want to try with my crash branch, which includes that patch, or wait till I've PRed it (I started PRing one of the deps).

**Comment by woodser** [comment]
I can try the branch to see if it resolve the issue.

**Comment by moneromooo-monero** [comment]
It converts your db to v6 btw. Use on a copy.

**Comment by moneromooo-monero** [comment]
https://github.com/monero-project/monero/pull/6616 has all the necessary patches for this without the other stuff.

**Comment by woodser** [comment]
Will #6616 update my db to v6?

**Comment by moneromooo-monero** [comment]
Yes.

**Comment by woodser** [comment]
Still seeing this issue with #6616.

**Comment by moneromooo-monero** [comment]
I'll try again with your wallet soon, in case I messed something up.

**Comment by woodser** [comment]
Apparently `get_reserve_proof` validates against the account's strict balance, which counts pending spent outputs and can exceed a single account's reported balance when unconfirmed txs exist. It verifies correctly with an amount greater than the wallet's overall balance.

**Comment by woodser** [comment]
AI's assessment of the core issue:

Both the balance guard and output selection in `get_reserve_proof` (`wallet2.cpp`)
use the **strict** spent-check:

- guard: `balance(account, true) < requested`
- selection: `!is_spent(td, true)`

`is_spent(td, true)` only treats an output as spent once `m_spent_height > 0` (i.e.
**confirmed** spent), so outputs already committed to a broadcast-but-unmined tx are
still counted as available and can be included in the proof. Their key images aren't
on-chain yet, so `check_reserve_proof` sees them as unspent — letting a wallet with
pending outgoing txs prove reserves it has already committed to spending.

**Fix direction:** guard/select against genuinely-unspent outputs (`!is_spent(td, false)`)
instead of the confirmed-spent-only strict view.

**Comment by selsta** [comment]
#11213
